lib/matplotlib/cbook.py

```diff
@@ -2428,6 +2428,107 @@ def safe_first_element(obj):
     return next(iter(obj))
 
 
+def normalize_kwargs(kw, alias_mapping=None, required=(), forbidden=(),
+                     allowed=None):
+    """Helper function to normalize kwarg inputs
+
+    The order they are resolved are:
+
+     1. aliasing
+     2. required
+     3. forbidden
+     4. allowed
+
+    This order means that only the canonical names need appear in
+    `allowed`, `forbidden`, `required `
+
+    Parameters
+    ----------
+
+    alias_mapping, dict, optional
+        A mapping between a canonical name to a list of
+        aliases, in order of precedence from lowest to highest.
+
+        If the canonical value is not in the list it is assumed to have
+        the highest priority.
+
+    required : iterable, optional
+        A tuple of fields that must be in kwargs.
+
+    forbidden : iterable, optional
+        A list of keys which may not be in kwargs
+
+    allowed : tuple, optional
+        A tuple of allowed fields.  If this non None, then raise if
+        `kw` contains any keys not in the union of `required`
+        and `allowed`.  To allow only the required fields pass in
+        ``()`` for `allowed`
+
+    Raises
+    ------
+    TypeError
+        To match what python raises if invalid args/kwargs are passed to
+        a callable.
+
+    """
+    # deal with default value of alias_mapping
+    alias_mapping = alias_mapping if alias_mapping is not None else {}
+
+    # make a local so we can pop
+    kw = dict(kw)
+    # output dictionary
+    ret = dict()
+
+    # hit all alias mappings
+    for canonical, alais_list in six.iteritems(alias_mapping):
+
+        # the alais lists are ordered from lowest to highest priority
+        # so we know to use the last value in this list
+        tmp = []
+
+        for a in alais_list:
+            try:
+                tmp.append(kw.pop(a))
+            except KeyError:
+                pass
+        # if canonical is not in the alais_list assume highest priority
+        if canonical not in alais_list:
+            try:
+                tmp.append(kw.pop(canonical))
+            except KeyError:
+                pass
+        # if we found anything in this set of aliases put it in the return
+        # dict
+        if tmp:
+            ret[canonical] = tmp[-1]
```

Comment:
My idea of what this would be used for is cleaning up kwargs dicts (so values we don't have defaults for) so something like

```python

def foo(x, y, **kwargs):
    kwargs = normalize_kwargs(kwargs, alias_mapping={'color': ('c',), 'linewidth': ('lw',)})
    return Line2D(x, y, **kwargs)
```

So

```python
foo(x, y, c='r') 
foo(x, y, color='green')
foo(x, y, c='red', color='green')  # warns, gives green
```

all do the right thing.